### PR TITLE
[fix] Remove trtllm ragged mla prefills

### PR DESCRIPTION
**Pull request by evezhier**
## Purpose
Removing output and workspace_buffer prefills for `run_prefill_context_chunk_trtllm_ragged`. This kernel reads and writes only tokens having history. Empty buffers may cause numeric issues for tokens w/o context. This PR modifies `merge_attn_states` to be aware of this.

## Test Plan
- [+]  `test_merge_attn_states` now accepts `prefills_with_context`
- [+] accuracy (`lm_eval`)
- [+] performance (`vllm bench serve`)

## Test Result
`VLLM_ATTENTION_BACKEND=FLASHINFER_MLA VLLM_USE_TRTLLM_RAGGED_DEEPSEEK_PREFILL=1 python -m vllm.entrypoints.openai.api_server --model="deepseek-ai/DeepSeek-V2-Lite-Chat" --no-enable-prefix-caching --seed 42`

### Accuracy: 

`lm_eval --model local-completions --model_args model=deepseek-ai/DeepSeek-V2-Lite-Chat,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=256 --tasks arc_challenge --num_fewshot 8 --batch_size 1 --seed 42`

**main**
|    Tasks    |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|-------------|------:|------|-----:|--------|---|-----:|---|-----:|
|arc_challenge|      1|none  |     8|acc     |↑  |0.5717|±  |0.0145|
|             |       |none  |     8|acc_norm|↑  |0.6049|±  |0.0143|

**This PR**
|    Tasks    |Version|Filter|n-shot| Metric |   |Value |   |Stderr|
|-------------|------:|------|-----:|--------|---|-----:|---|-----:|
|arc_challenge|      1|none  |     8|acc     |↑  |0.5717|±  |0.0145|
|             |       |none  |     8|acc_norm|↑  |0.6067|±  |0.0143|

### Performance:

`vllm bench serve --model deepseek-ai/DeepSeek-V2-Lite-Chat --num-prompts 2048 --trust-remote-code --ignore-eos --max-concurrency 2048 --random-input-len 2048 --random-output-len 512 --seed 42`

**main**
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             2048      
Benchmark duration (s):                  72.23     
Total input tokens:                      4192256   
Total generated tokens:                  1048576   
Request throughput (req/s):              28.35     
Output token throughput (tok/s):         14516.47  
Peak output token throughput (tok/s):    26880.00  
Peak concurrent requests:                2048.00   
Total token throughput (tok/s):          72554.02  
---------------Time to First Token----------------
Mean TTFT (ms):                          27606.91  
Median TTFT (ms):                        27032.76  
P99 TTFT (ms):                           55005.85  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          55.42     
Median TPOT (ms):                        59.52     
P99 TPOT (ms):                           63.07     
---------------Inter-token Latency----------------
Mean ITL (ms):                           55.42     
Median ITL (ms):                         40.45     
P99 ITL (ms):                            90.35     

**This PR**
============ Serving Benchmark Result ============
Successful requests:                     2048      
Failed requests:                         0         
Maximum request concurrency:             2048      
Benchmark duration (s):                  70.16     
Total input tokens:                      4192256   
Total generated tokens:                  1048576   
Request throughput (req/s):              29.19     
Output token throughput (tok/s):         14945.22  
Peak output token throughput (tok/s):    27004.00  
Peak concurrent requests:                2048.00   
Total token throughput (tok/s):          74696.90  
---------------Time to First Token----------------
Mean TTFT (ms):                          26826.12  
Median TTFT (ms):                        26223.21  
P99 TTFT (ms):                           53445.80  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          53.82     
Median TPOT (ms):                        57.81     
P99 TPOT (ms):                           61.01     
---------------Inter-token Latency----------------
Mean ITL (ms):                           53.82     
Median ITL (ms):                         39.76     
P99 ITL (ms):                            86.36     

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 1]
```diff
@@ -2378,14 +2381,13 @@ def _run_prefill_context_chunk_trtllm_ragged(
         assert prefill.chunked_context.seq_lens[chunk_idx] is not None
         assert prefill.workspace_buffer is not None
 
-        out = torch.zeros(
+        out = torch.empty(
```
[the remark this answers is not in the document]
Comment (evezhier) [reply vllm/model_executor/layers/attention/mla_attention.py:2396]: This is a precise description of why this task appeared, thanks, Gemini :) merge_attn_states has been modified to handle this.

File: vllm/v1/attention/ops/merge_attn_states.py [hunk 2]
```diff
@@ -13,6 +13,7 @@ def merge_attn_states(
     suffix_output: torch.Tensor,
     suffix_lse: torch.Tensor,
     output_lse: torch.Tensor | None = None,
+    prefill_tokens_with_context: int | None = None,
 ) -> None:
     # NOTE(DefTruth): Currently, custom merge_attn_states CUDA kernel
```
Comment (pavanimajety) [inline vllm/v1/attention/ops/merge_attn_states.py:46]: Nit: Could we add a doc string here explaining all the arguments, including the new one? Especially that what it means when it is None vs when it is an integer.

**Review by pavanimajety — COMMENTED** [review]
Thanks for the PR, do we see any reduced perf due to the additional logic for batches of requests with no history context?

File: vllm/v1/attention/ops/merge_attn_states.py (hunk 2 above)
Comment (evezhier) [reply vllm/v1/attention/ops/merge_attn_states.py:46]: Done

File: vllm/model_executor/layers/attention/mla_attention.py [hunk 3]
```diff
@@ -1864,6 +1865,11 @@ def build(
                     if self._use_cudnn_prefill
                     else MLACommonPrefillMetadata.ChunkedContextMetadata
                 )
+                prefill_tokens_with_context = None
+                if num_prefills_with_context_cpu > 0:
+                    prefill_tokens_with_context = prefill_query_start_loc[
+                        num_prefills_with_context_cpu
+                    ].item()
```
Comment (LucasWilkinson) [inline vllm/model_executor/layers/attention/mla_attention.py:1875]: we should prefer `prefill_query_start_loc_cpu` over the device side tensor if doing `.item()` to avoid D->H sync

Comment (LucasWilkinson) [reply vllm/model_executor/layers/attention/mla_attention.py:1875]: I think you may actually want to consider using `split_decodes_prefills_and_extends` instead of `split_decodes_and_prefills` to achieve this

**Review by LucasWilkinson — APPROVED** [review]
Overall LGTM; thanks for doing this! left one comment thats worth addressing

File: vllm/model_executor/layers/attention/mla_attention.py (hunk 3 above)
Comment (evezhier) [reply vllm/model_executor/layers/attention/mla_attention.py:1875]: There is no `prefill_query_start_loc_cpu`, to create it I had to duplicate the operation in the `MLACommonMetadataBuilder`:
```
prefill_query_start_loc = (
        query_start_loc[reqs_start:] - query_start_loc[reqs_start]
)
prefill_query_start_loc_cpu = (
        query_start_loc_cpu[reqs_start:] - query_start_loc_cpu[reqs_start]
)
```
It's awkward but it saves a little perf, thanks.
